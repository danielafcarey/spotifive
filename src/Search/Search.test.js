import React from 'react';
import { shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import {
  Search,
  mapStateToProps,
  mapDispatchToProps
} from './Search';


describe('Search', () => {
  let mockProps;
  let wrapper;

  beforeEach(() => {
    mockProps = {
      accessToken: 'a',
      user: { 
        userInfo: {userId: 1 },
        loggedIn: true
      },
      searchResults: { 
        searchResults: [] 
      },
      artist: { 
        artist: { artistId: 1 },
        artistError: null
      },
      submitUpdateSearch: jest.fn(),
      submitUpdateArtist: jest.fn(),
      updateArtist: jest.fn(),
      history: { push: jest.fn() }
    };
    wrapper = shallow(<Search { ...mockProps } />);
  })

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('has a default state of searchInput', () => {
    expect(wrapper.state('searchInput')).toEqual('');
  })
  
  it('matches the snapshot if user is not logged in', () => {
    mockProps.user.loggedIn = false;
    const wrapper = shallow(<Search { ...mockProps } />);

    expect(wrapper).toMatchSnapshot()
  })

  it('matches the snapshot if there are search results', () => {
    mockProps.searchResults = { 
      searchResults: [{ name: 'hi', id: 1 }] 
    };
    const wrapper = shallow(<Search { ...mockProps } />);

    expect(wrapper).toMatchSnapshot();
  })

  it('matches the snapshot if there is no artist selected', () => {
    mockProps.artist.artist = {};
    const wrapper = shallow(<Search { ...mockProps } />);

    expect(wrapper).toMatchSnapshot();
  })

  describe('handleChange', () => {

    it('sets the state of searchInput', () => {
      const mockEvent = { target: { value: 'some input' } };
      const expected = 'some input';
      
      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state('searchInput')).toEqual(expected);
    })

  })

  describe('handleSubmit', () => {
    let mockEvent;

    beforeEach(() => {
      mockEvent = { preventDefault: () => {} }
    })

    it('calls props.submitUpdateSearch with the correct arguments', () => {
      wrapper.setState({ searchInput: 'hi' })
      const wrapperInst = wrapper.instance();
      wrapperInst.handleSubmit(mockEvent);

      expect(wrapperInst.props.submitUpdateSearch).toHaveBeenCalledWith('hi', 'a');
    })

    it('resets the state to an empty string', () => {
      wrapper.instance().handleSubmit;

      expect(wrapper.state('searchInput')).toEqual('');
    })

  })

  describe('selectArtist', () => {
    let mockId;

    beforeEach(() => {
      mockId = 1
    })

    it('calls props.submitUpdateArtist with the correct arguments', () => {
      wrapper.instance().selectArtist('1');

      expect(mockProps.submitUpdateArtist).toHaveBeenCalledWith('1', 'a')
    })

  })

  describe('mapStateToProps', () => {

    it('returns a state object with accessToken, user, and artist', () => {
      const state = {
        accessToken: 1,
        user: { userId: 1 },
        artist: { artistId: 1 },
        searchResults: ['artist1', 'artist2'],
        fakeProp: 'do not add me'
      }
      const expected = {
        accessToken: 1,
        user: { userId: 1 },
        searchResults: ['artist1', 'artist2'],
        artist: { artistId: 1 }
      }
      const result = mapStateToProps(state);

      expect(result).toEqual(expected);
    })

  })

  describe('mapDispatchToProps', () => {
    
    it('calls dispatch with the correct arguments when submitUpdateSearch is invoked', () => {
      const dispatch = jest.fn();
      const mappedProps = mapDispatchToProps(dispatch);
      const mockAction = {
        type: 'SUBMIT_UPDATE_SEARCH',
        searchString: 'hi',
        accessToken: 'me'
      }
      mappedProps.submitUpdateSearch('hi', 'me');

      expect(dispatch).toHaveBeenCalledWith(mockAction)
    })

    it('calls dispatch with the correct arguments when submitUpdateArtist is invoked', () => {
      const dispatch = jest.fn();
      const mappedProps = mapDispatchToProps(dispatch);
      const mockAction = {
        type: 'SUBMIT_UPDATE_ARTIST',
        artistId: 'hi',
        accessToken: '2'
      }
      mappedProps.submitUpdateArtist('hi', '2');

      expect(dispatch).toHaveBeenCalledWith(mockAction)
    })

    it('calls dispatch with the correct arguments when submitUpdateArtist is invoked', () => {
      const dispatch = jest.fn();
      const mappedProps = mapDispatchToProps(dispatch);
      const mockAction = {
        type: 'UPDATE_ARTIST',
        artist: 'i am an artist'
      }
      mappedProps.updateArtist('i am an artist');

      expect(dispatch).toHaveBeenCalledWith(mockAction)
    })

  })

})
