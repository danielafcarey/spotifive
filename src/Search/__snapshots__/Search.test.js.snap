// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    accessToken="a"
    artist={
        Object {
            "artist": Object {
              "artistId": "1",
            },
            "artistError": null,
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    searchResults={
        Object {
            "searchError": null,
            "searchResults": Array [],
          }
    }
    submitUpdateArtist={[Function]}
    submitUpdateSearch={[Function]}
    updateArtist={[Function]}
    user={
        Object {
            "loggedIn": true,
            "userInfo": Object {
              "userId": "1",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
>
          <input
                    onChange={[Function]}
                    placeholder="Search for an artist"
                    value=""
          />
          <p
                    className="search-error"
          />
          <button>
                    Search
          </button>
</form>,
        <Instructions />,
      ],
      "className": "search",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              placeholder="Search for an artist"
              value=""
/>,
            <p
              className="search-error"
/>,
            <button>
              Search
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Search for an artist",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "className": "search-error",
            },
            "ref": null,
            "rendered": "",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
>
            <input
                        onChange={[Function]}
                        placeholder="Search for an artist"
                        value=""
            />
            <p
                        className="search-error"
            />
            <button>
                        Search
            </button>
</form>,
          <Instructions />,
        ],
        "className": "search",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                placeholder="Search for an artist"
                value=""
/>,
              <p
                className="search-error"
/>,
              <button>
                Search
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Search for an artist",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "search-error",
              },
              "ref": null,
              "rendered": "",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Search matches the snapshot if there are search results 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    accessToken="a"
    artist={
        Object {
            "artist": Object {
              "artistId": "1",
            },
            "artistError": null,
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    searchResults={
        Object {
            "searchResults": Array [
              Object {
                "id": 1,
                "name": "hi",
              },
            ],
          }
    }
    submitUpdateArtist={[Function]}
    submitUpdateSearch={[Function]}
    updateArtist={[Function]}
    user={
        Object {
            "loggedIn": true,
            "userInfo": Object {
              "userId": "1",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
>
          <input
                    onChange={[Function]}
                    placeholder="Search for an artist"
                    value=""
          />
          <p
                    className="search-error"
          />
          <button>
                    Search
          </button>
</form>,
        <SearchResults
          searchResults={
                    Object {
                              "searchResults": Array [
                                Object {
                                  "id": 1,
                                  "name": "hi",
                                },
                              ],
                            }
          }
          selectArtist={[Function]}
/>,
      ],
      "className": "search",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              placeholder="Search for an artist"
              value=""
/>,
            <p
              className="search-error"
/>,
            <button>
              Search
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Search for an artist",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "className": "search-error",
            },
            "ref": null,
            "rendered": "",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "searchResults": Object {
            "searchResults": Array [
              Object {
                "id": 1,
                "name": "hi",
              },
            ],
          },
          "selectArtist": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
>
            <input
                        onChange={[Function]}
                        placeholder="Search for an artist"
                        value=""
            />
            <p
                        className="search-error"
            />
            <button>
                        Search
            </button>
</form>,
          <SearchResults
            searchResults={
                        Object {
                                    "searchResults": Array [
                                      Object {
                                        "id": 1,
                                        "name": "hi",
                                      },
                                    ],
                                  }
            }
            selectArtist={[Function]}
/>,
        ],
        "className": "search",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                placeholder="Search for an artist"
                value=""
/>,
              <p
                className="search-error"
/>,
              <button>
                Search
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Search for an artist",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "search-error",
              },
              "ref": null,
              "rendered": "",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "searchResults": Object {
              "searchResults": Array [
                Object {
                  "id": 1,
                  "name": "hi",
                },
              ],
            },
            "selectArtist": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Search matches the snapshot if there is no artist selected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    accessToken="a"
    artist={
        Object {
            "artist": Object {},
            "artistError": null,
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    searchResults={
        Object {
            "searchError": null,
            "searchResults": Array [],
          }
    }
    submitUpdateArtist={[Function]}
    submitUpdateSearch={[Function]}
    updateArtist={[Function]}
    user={
        Object {
            "loggedIn": true,
            "userInfo": Object {
              "userId": "1",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
>
          <input
                    onChange={[Function]}
                    placeholder="Search for an artist"
                    value=""
          />
          <p
                    className="search-error"
          />
          <button>
                    Search
          </button>
</form>,
        <Instructions />,
      ],
      "className": "search",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              placeholder="Search for an artist"
              value=""
/>,
            <p
              className="search-error"
/>,
            <button>
              Search
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Search for an artist",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "className": "search-error",
            },
            "ref": null,
            "rendered": "",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
>
            <input
                        onChange={[Function]}
                        placeholder="Search for an artist"
                        value=""
            />
            <p
                        className="search-error"
            />
            <button>
                        Search
            </button>
</form>,
          <Instructions />,
        ],
        "className": "search",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                placeholder="Search for an artist"
                value=""
/>,
              <p
                className="search-error"
/>,
              <button>
                Search
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Search for an artist",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "search-error",
              },
              "ref": null,
              "rendered": "",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Search matches the snapshot if user is not logged in 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    accessToken="a"
    artist={
        Object {
            "artist": Object {
              "artistId": "1",
            },
            "artistError": null,
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    searchResults={
        Object {
            "searchError": null,
            "searchResults": Array [],
          }
    }
    submitUpdateArtist={[Function]}
    submitUpdateSearch={[Function]}
    updateArtist={[Function]}
    user={
        Object {
            "loggedIn": false,
            "userInfo": Object {
              "userId": "1",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "push": false,
      "to": "/",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "push": false,
        "to": "/",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
